Copyediting, part 2 (#240)
chef/effortless: effortless-chef-io/effortless-config.md

## Patterns

### Chef Repo Cookbook Pattern

> This pattern builds Policyfiles using the chef-repo pattern for organizing your cookbooks. More information about the [chef-repo](https://docs.chef.io/chef_repo/) pattern can be found in our Chef Infra Documentation. View an example of this pattern [here](https://github.com/chef/effortless/tree/master/examples/effortless_config/chef_repo_pattern).

Reviewer: ```suggestion
This pattern builds Policyfiles using the [chef-repo pattern](https://github.com/chef/effortless/tree/master/examples/effortless_config/chef_repo_pattern) for organizing your cookbooks. Our Chef Infra documentation contains more information about the [chef-repo](https://docs.chef.io/chef_repo/) pattern.
```

or something else that doesn't have a here link

Reviewer 2: Okay that makes sense.

Reviewer 3: ```suggestion
This pattern uses the [chef-repo](https://docs.chef.io/chef_repo/) to store and organize everything you need to define your infrastructure with Chef Infra, including:

- Cookbooks (including recipes, attributes, custom resources, libraries, and templates)
- Data bags
- Policyfiles

The Effortless Config pattern builds Policyfiles using the chef-repo. View an [example chef-repo](https://github.com/chef/effortless/tree/master/examples/effortless_config/chef_repo_pattern) in the Chef Effortless GitHub repository.
```

1. To use this pattern, navigate to the chef-repo directory that you want to use:

   ```bash
   cd chef-repo
   ```

1. Create a `habitat` directory from the command line with:

   ```bash
   mkdir habitat
   ```

1. Make a plan file

>    Use a `plan.ps1` for a cookbook targeting Windows. Use a `plan.sh` for a cookbook targeting Linux. If the cookbook targets both Windows and Linux, you can have both a `plan.ps1` and a `plan.sh` in the `habitat` directory. Create a plan in Linux with the following command:

Reviewer 3: ```suggestion
   Make a `plan.ps1` for a cookbook that targets Microsoft Windows systems and a `plan.sh` for a cookbook that targets Linux systems. You can have both a `plan.ps1` and a `plan.sh` in the `habitat` directory for cookbooks that target both systems. Create a plan in Linux with the following command:
```

   ```bash
   touch plan.sh
   ```

1. Add information about the cookbook to the plan

   Add this information to the `plan.sh` file:

   ```bash
   pkg_name=<NAME FOR MY POLICYFILE>
   pkg_origin=<MY ORIGIN>
   pkg_version="0.1.0"
   pkg_maintainer="MY NAME AND EMAIL"
   pkg_license=("Apache-2.0")
   pkg_scaffolding="chef/scaffolding-chef-infra"
   pkg_svc_user=("root")
   scaffold_policy_name="<MY POLICYFILE NAME>"
   ```

> 1. Create a Policyfile directory in the chef-repo and build a Policyfile

Author: @TheLunaticScripter Is there a specific command required to build a Policyfile?

Reviewer 2: I totally didn't realize but it looks like `chef generate policyfile`  creates a `Policyfile.rb` in the directory that the command is run in.

Author: Oh cool! Thanks for that! So in "building" a Policyfile, do you mean that you're adding content to the empty Policyfile?

Reviewer 2: Yes createing the file and adding the content.

Reviewer 4: We could reference the creation if we wanted as well:

```shell
mkdir policyfiles
chef generate policyfile policyfiles/MyPolicyfile.rb                                                                                
Recipe: code_generator::policyfile
  * template[/workspace/my_repo/policyfiles/MyPolicyfile.rb] action create
    - create new file /workspace/my_repo/policyfiles/MyPolicyfile.rb
    - update content in file /workspace/my_repo/policyfiles/MyPolicyfile.rb from none to fc117f
    (diff output suppressed by config)
```

Reviewer 3: ```suggestion
1. If you do not have a policyfiles directory in your chef-repo, create one:
 
   ```
   mkdir policyfiles
   ```

1. Generate a Policyfile

  ```
  chef generate policyfile policyfiles/my_policyfile
  ```
```

Reviewer 3: Does chef generate `my_policyfile` require the `.rb` extension?

Reviewer 2: it does not require the `.rb` extension

- Cookbooks (including recipes, attributes, custom resources, libraries, and templates)
- Data bags
- Policyfiles
  
The Chef Effortless GitHub repository has an [example chef-repo](https://github.com/chef/effortless/tree/master/examples/effortless_config/chef_repo_pattern) for you to explore.

1. To use this pattern, navigate to the chef-repo directory that you want to use:

   ```bash
   cd chef-repo
   ```

1. Create a `habitat` directory from the command line with:

   ```bash
   mkdir habitat
   ```

1. Make a plan file

   Make  a `plan.ps1` for a cookbook that targets Microsoft Windows systems and a `plan.sh` for a cookbook that targets Linux systems. You can have both a `plan.ps1` and a `plan.sh` in the `habitat` directory for cookbooks that target both systems. Create a Linux plan:

   ```bash
   touch plan.sh
   ```

1. Add information about the cookbook to the plan

   Add this information to the `plan.sh` file:

   ```bash
   pkg_name=<NAME FOR MY POLICYFILE>
   pkg_origin=<MY ORIGIN>
   pkg_version="0.1.0"
   pkg_maintainer="MY NAME AND EMAIL"
   pkg_license=("Apache-2.0")
   pkg_scaffolding="chef/scaffolding-chef-infra"
   pkg_svc_user=("root")

>    scaffold_policy_name="<MY POLICYFILE NAME>"

Author: @kagarmoe What about this one? :)

Author: There's another on line 43

Reviewer 3: Good catch! `my_policyfile` because we can't have spaces in the policyfile name

Author: And it goes with the example you put further down. BRB gonna update the other PR, too.

   suites:
     - name: base
       provisioner:

>          arguments: ["<MY ORIGIN>", "<MY PACKAGE NAME>"]

Reviewer 3: ```suggestion
         arguments: ["<MY_ORIGIN>", "<MY_PACKAGE_NAME>"]
```

   echo "Installing $latest_hart_file"
   hab pkg install $latest_hart_file

>    echo "Determining pkg_prefix for $latest_hart_file"

Reviewer 3: @TheLunaticScripter Can we check it this typo is in the code?

Reviewer 2: yep I'll take a look.

   cd $pkg_prefix
   hab pkg exec $pkg_origin/$pkg_name chef-client -z -c $pkg_prefix/config/bootstrap-config.rb
   ```

1. Run Test Kitchen to ensure the cookbook works

   Use this command to spin up a CentOS 7 virtual machine (VM) locally and run the cookbook using the latest Chef Infra Client:

   ```bash
   kitchen converge base-centos
   ```

   If you experience errors in this Chef run, you may need to supply attributes or make modifications to your Policyfile, so that your cookbook can run using the latest Chef Infra Client.

1. When ready, delete the VM instance by running:

   ```bash
   kitchen destroy
   ```

1. Upload the Policyfile pkg to Chef Habitat builder by running the following commands:

   ```bash
   source results/lastbuild.env
   hab pkg upload results/$pkg_artifact
   ```

1. To run the Policyfile on a system, install Chef Habitat services and run:

   ```bash
   hab svc load <my_origin>/<my_policyfile_name>
   ```

> ### Policyfile Cookbook pattern

Reviewer 3: ```suggestion
### Policyfile Cookbook Pattern
```

> This pattern builds an artifact for a single cookbook. View an example of this pattern [here](https://github.com/chef/effortless/tree/master/examples/effortless_config/policyfile_cookbook_pattern).

Reviewer: ```suggestion
The [cookbook pattern]((https://github.com/chef/effortless/tree/master/examples/effortless_config/policyfile_cookbook_pattern)) builds an artifact for a single cookbook.
```

Reviewer 3: We need to be clear that we're providing an example Policyfile cookbook. Cookbooks vary.

Reviewer 3: ```suggestion
This pattern builds a Chef Habitat artifact for the Policyfile cookbook. You can find an [example Policyfile cookbook](https://github.com/chef/effortless/tree/master/examples/effortless_config/policyfile_cookbook_pattern) in the Chef Effortless GitHub repository.
```

   ```bash
   hab svc load <my_origin>/<my_policyfile_name>
   ```

### Policyfile Cookbook Pattern

This pattern builds a Chef Habitat artifact for the Policyfile cookbook. You can find an [example Policyfile cookbook](https://github.com/chef/effortless/tree/master/examples/effortless_config/policyfile_cookbook_pattern) in the Chef Effortless GitHub repository.

1. To use this pattern, navigate to the cookbook you want to use:

   ```bash
   cd chef-repo/cookbooks/foo_cookbook
   ```

1. Create a `habitat` directory from the command line with:

   ```bash
   mkdir habitat
   ```

1. Make a plan file

   Use a `plan.ps1` for a cookbook targeting Windows. Use a `plan.sh` for a cookbook targeting Linux. If the cookbook targets both Windows and Linux, you can have both a `plan.ps1` and a `plan.sh` in the `habitat` directory. Create a plan in Linux with the following command:

   ```bash
   touch plan.sh
   ```

1. Add some information about the cookbook to the plan

   plan.sh

   ```bash
   pkg_name=<Name of my cookbook artifact>
   pkg_origin=<My Origin>
   pkg_version="<Cookbook version>"
   pkg_maintainer="<My Name>"

>    pkg_license=("<License for my cookbook example Apache-2.0>")

Reviewer 3: ```suggestion
   pkg_license=("<License for my_cookbook example Apache-2.0>")
```

   # Specify a custom source for a single cookbook:
   cookbook '<Your Cookbook Name>', path: '.'
   ```

1. Build the package:

   ```bash

>    hab pkg build <My Cookbook Name>

Reviewer 3: ```suggestion
   hab pkg build <my_cookbook>
```

   suites:
   - name: base
      provisioner:

>          arguments: ["<MY ORIGIN>", "<MY COOKBOOK NAME>"]

Reviewer 3: ```suggestion
         arguments: ["<MY_ORIGIN>", "<MY_COOKBOOK_NAME>"]
```

   suites:
   - name: base
      provisioner:

>          arguments: ["<MY_ORIGIN>", "<MY_COOKBOOK_NAME>"]

Reviewer 3: ```suggestion
         arguments: ["<my_origin>", "<my_cookbook>"]
```
